Add unit tests in Clarity

friedger opened the pull request:
### Description
This PR adds the ability to write clarinet-sdk tests in Clarity.

As an example a simple unit tests and a simple flow tests was added. These tests do not test anything meaningful, but show how these tests work.

### Applicable issues
- fixes #3976 

### Additional info (benefits, drawbacks, caveats)

### Checklist
- [ ] Test coverage for new or modified code paths
- [ ] Changelog is updated
- [ ] Required documentation changes (e.g., `docs/rpc/openapi.yaml` and `rpc-endpoints.md` for v2 endpoints, `event-dispatcher.md` for new events)
- [ ] New clarity functions have corresponding PR in `clarity-benchmarking` repo
- [ ] New integration test(s) added to `bitcoin-tests.yml`

friedger committed:
feat: add unit tests in clarity

friedger committed:
chore: add log file for flow tests

friedger committed:
chore: push package-lock.json

moodmosaic reviewed (commented):
LGTM

MarvinJanssen reviewed (approved):
Naturally I much prefer writing tests in Clarity. The question is do the other code owners like (or at least are they OK with) bringing this new testing paradigm into the repository. Who decides?

moodmosaic commented on contrib/core-contract-tests/contracts/bns-tests/bns_flow_test.clar:
```diff
@@ -0,0 +1,19 @@
+(define-constant ERR_NAMESPACE_NOT_FOUND 1005)
+
+;; @name: test delegation to wallet_2, stacking and revoking
+(define-public (test-name-registration)
```
If we can sort this part out, I can integrate this with fast-check and we'll have it generate/fuzz everything for us. What's very nice about this is the shrinking part; if there's a bug/edge-case detected, fast-check will give us the minimum counterexample and a seed to reproduce the failure (if the bug/edge-case is detected when running on CI).

jcnelson commented on contrib/core-contract-tests/tests/clar/clar-flow.test.ts:
```diff
@@ -0,0 +1,138 @@
+import { ParsedTransactionResult, tx } from "@hirosystems/clarinet-sdk";
```
Hey, do you think you can add some method-level comments for each of these functions in this file and the `clar.test.ts` file?  I'm having a hard time following how these functions fit together.  Thanks!

MarvinJanssen replied:
This parser evolved over time during our work on sBTC Mini. We just discussed in the Clarity WG to clean it up and perhaps introduce a more robust parser. The current files have served our needs and were recently edited to support `clarinet-sdk`.

friedger replied:
Comments have been added

jcnelson commented on contrib/core-contract-tests/tests/clar/utils/clarity-parser.ts:
```diff
@@ -0,0 +1,346 @@
+
+export function getContractName(contractId: string) {
+  return contractId.split(".")[1];
+}
```
This parser looks somewhat involved.  Is there a way to add a (simple) set of unit tests to verify that all the test parsers work as intended?  Thanks!

moodmosaic replied:
@friedger @jcnelson happy to do this with fast-check? This way we’ll test and have the ability to also fuzz the interpreter at the same time.

friedger replied:
Code improvements have been pushed

moodmosaic replied:
FYI, some prop/fuzz tests I added can be found in https://github.com/stacks-network/stacks-core/pull/4100/commits/3e9b3af440ade34268b7e5c716ad0534c3982759 and https://github.com/stacks-network/stacks-core/pull/4100/commits/34209043ee101bb5923d40108a5f60aeb6f61962.

friedger committed:
chore: add comments

friedger committed:
chore: improve code readability

friedger committed:
chore: add more tests

friedger committed:
Merge branch 'next' into feat/clarinet-flow-tests

moodmosaic committed:
feat(tests): Add clarity-parser-string-to-cv.prop.test.ts property-based tests

- Implemented property-based testing for string, uint, and tuple conversions using fast-check.
- Defined custom arbitraries for generating 128-bit unsigned (uint128) and signed (int128) integers.
- Enhanced test robustness and coverage by automatically testing a wide range of input values.
- Included fast-check's shrinking feature to identify minimal failure cases, aiding in effective debugging.

moodmosaic committed:
feat(tests): Add clarity-parser.prop.test.ts property-based tests

friedger committed:
Merge branch 'next' into feat/clarinet-flow-tests

setzeus reviewed (approved):
Pulled & followed the logic due to familiarity w/ mini. My only feedback here is to make an issue for documenting how these tests are built & how a contributor can write additional tests.

LGTM.

friedger committed:
Merge branch 'next' into feat/clarinet-flow-tests

wileyj commented:
@MarvinJanssen @friedger I like the approach here, and what you're trying to do. 
however, i think it might be better to think about these changes as something that would be easier to maintain as a separate repo entirely (possibly even a monorepo with a few other testing repos that i know of). 

my concern is that if we need to make changes, the bar to getting them merged to `master` is much higher in this repo vs a separate repo. 

merging to next, and updating next is less of an issue i'll agree. i'm thinking longer-term though, when eventually next is merged to master and we may find issues/improvement that should be made. having this in a separate repo would make those changes much easier to make.

friedger committed:
Merge branch 'next' into feat/clarinet-flow-tests

friedger commented:
@wileyj I agree. 

As there is currently no other repo, this seems to be the only solution to get a good discussion. 

Merging a version update of the tool into main should be as hard as updating the source code 😉

wileyj commented:
> @wileyj I agree.
> 
> As there is currently no other repo, this seems to be the only solution to get a good discussion.
> 
> Merging a version update of the tool into main should be as hard as updating the source code 😉

that's what concerns me actually - unless i'm misunderstanding the reasoning behind this PR, i see a lot of benefit in moving most (or all) of the tooling outside of the blockchain repo (we can still maintain a higher bar for merging PR's, but maybe not quite as high as the blockchain repo requires). 
the tests themselves, i have no opinion on - i'm mostly looking at the tooling code/dirs in this PR. 
we can discuss wednesday, but the goal for me is to make updating the tooling as easy as it needs to be.

obycode commented:
Being discussed tomorrow.

friedger commented:
using clarunit now: https://github.com/stacks-network/stacks-core/issues/4203#issuecomment-1870391449

blockstack-devops commented:
This pull request has been automatically locked since there has not been any recent activity after it was closed. Please open a new issue for related bugs.
